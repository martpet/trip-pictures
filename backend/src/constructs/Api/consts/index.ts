export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
