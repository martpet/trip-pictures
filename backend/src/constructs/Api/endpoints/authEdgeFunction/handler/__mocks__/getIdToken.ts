export const getIdToken = vi
  .fn()
  .mockName('getIdToken')
  .mockResolvedValue('dummyIdToken');
