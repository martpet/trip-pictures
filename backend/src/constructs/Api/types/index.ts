export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
