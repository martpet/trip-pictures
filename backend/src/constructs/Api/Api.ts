import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
