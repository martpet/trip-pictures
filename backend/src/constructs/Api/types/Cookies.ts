import { ConditionalKeys } from 'type-fest';

import { apiOptions } from '~/consts';
import { ApiRouteOptions } from '~/types';

export type PathWithCookies = ConditionalKeys<
  typeof apiOptions,
  Pick<ApiRouteOptions, 'cookies'>
>;

export type Cookies<T extends PathWithCookies> = Partial<
  Record<(typeof apiOptions)[T]['cookies'][number], string>
>;

export type AllCookies = Cookies<PathWithCookies>;
