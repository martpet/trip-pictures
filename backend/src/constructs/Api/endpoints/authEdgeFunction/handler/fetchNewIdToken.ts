import fetch from 'node-fetch';

import { FetchNewIdTokenResponse } from '~/constructs/Api/types';
import { authPaths } from '~/consts';

import { updateSession } from './updateSession';

type FetchNewIdTokenProps = {
  refreshToken: string;
  sessionId: string;
  clientId: string;
};

export const fetchNewIdToken = async ({
  refreshToken,
  sessionId,
  clientId,
}: FetchNewIdTokenProps) => {
  const { authDomain } = globalAuthEdgeFunctionProps;

  console.info('Fetching new id token...');

  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'refresh_token',
      client_id: clientId,
      refresh_token: refreshToken,
    }),
  });

  const data = (await response.json()) as FetchNewIdTokenResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  await updateSession({ sessionId, idToken: data.id_token });

  console.info('New id token fetched and saved');

  return data.id_token;
};
