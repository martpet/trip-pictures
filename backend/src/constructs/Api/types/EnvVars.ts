import { SetRequired } from 'type-fest';

import { apiOptions } from '~/consts';
import { ApiMethodOptions, DeepValues } from '~/types';

type A = typeof apiOptions;

export type EnvVars<
  R extends keyof A,
  M extends keyof A[R]['methods'],
  O = A[R]['methods'][M]
> = O extends SetRequired<ApiMethodOptions, 'envVars'>
  ? Partial<Record<O['envVars'][number], string>>
  : {};

export type AllEnvVars = Record<DeepValues<A, 'envVars'>[number], string>;
