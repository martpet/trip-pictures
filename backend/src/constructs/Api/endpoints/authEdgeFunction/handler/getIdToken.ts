import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

import { getIdTokenPayload } from '~/constructs/Api/utils/getIdTokenPayload';
import { refreshTokenExpiredErrorMessage, region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

import { fetchNewIdToken } from './fetchNewIdToken';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const getIdToken = async (sessionId: string) => {
  const getCommand = new GetCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
  });

  const { Item } = await ddbDocClient.send(getCommand);
  const sessionsItem = Item as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  const { idToken, refreshToken, refreshTokenExpires } = sessionsItem;
  const { exp: idTokenExpires, aud: clientId } = await getIdTokenPayload(idToken);

  if ((idTokenExpires - 5) * 1000 > Date.now()) {
    return idToken;
  }

  if (refreshTokenExpires * 1000 > Date.now()) {
    return fetchNewIdToken({ refreshToken, sessionId, clientId });
  }

  throw new Error(refreshTokenExpiredErrorMessage);
};
