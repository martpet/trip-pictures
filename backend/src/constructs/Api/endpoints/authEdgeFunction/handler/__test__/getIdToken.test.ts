import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import millis from 'milliseconds';

import { IdTokenPayload } from '~/constructs/Api/types';
import {
  getIdTokenPayload,
  itCalls,
  itRejects,
  itResolves,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';

import { fetchNewIdToken } from '../fetchNewIdToken';
import { getIdToken } from '../getIdToken';

vi.mock('../fetchNewIdToken');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');

const ddbMock = mockClient(DynamoDBDocumentClient);
const originalDate = Date.now();
const idTokenExpires = (Date.now() + millis.days(1)) / 1000;
const refreshTokenExpires = (Date.now() + millis.years(1)) / 1000;

const args = ['dummySessionId'] as Parameters<typeof getIdToken>;

vi.mocked(fetchNewIdToken).mockResolvedValue('newDummyIdToken');

vi.mocked(getIdTokenPayload).mockResolvedValue({
  exp: idTokenExpires,
  aud: 'dummyAud',
} as IdTokenPayload);

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(GetCommand).resolves({
    Item: {
      idToken: 'dummyIdToken',
      refreshToken: 'dummyRefreshToken',
      refreshTokenExpires,
    },
  });
});

describe('getTokens', () => {
  itSendsAwsCommand(GetCommand, ddbMock, getIdToken, args);
  itResolves(getIdToken, args);

  describe('when "Item" is missing from "GetCommand" output', () => {
    beforeEach(() => {
      ddbMock.on(GetCommand).resolves({});
    });
    itRejects(getIdToken, args);
  });

  describe('when "idToken" has expired', () => {
    beforeAll(() => {
      vi.setSystemTime(new Date(idTokenExpires * 1000 + millis.days(1)));
    });
    afterAll(() => {
      vi.setSystemTime(originalDate);
    });
    itCalls(fetchNewIdToken, getIdToken, args);
    itResolves(getIdToken, args);
  });

  describe('when "refreshToken" has expired', () => {
    beforeAll(() => {
      vi.setSystemTime(new Date(refreshTokenExpires * 1000 + millis.days(1)));
    });
    afterAll(() => {
      vi.setSystemTime(originalDate);
    });
    itRejects(getIdToken, args);
  });
});
