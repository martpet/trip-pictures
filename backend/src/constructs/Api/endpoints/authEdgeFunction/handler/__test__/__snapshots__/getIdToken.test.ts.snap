// Vitest Snapshot v1

exports[`getTokens > resolves with a correct value 1`] = `"dummyIdToken"`;

exports[`getTokens > sends "GetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "Key": {
    "id": "dummySessionId",
  },
  "TableName": "TP-sessions",
}
`;

exports[`getTokens > when "Item" is missing from "GetCommand" output > rejects with a correct value 1`] = `[Error: could not find session with id "dummySessionId"]`;

exports[`getTokens > when "idToken" has expired > calls "fetchNewIdToken" with correct args 1`] = `
[
  [
    {
      "clientId": "dummyAud",
      "refreshToken": "dummyRefreshToken",
      "sessionId": "dummySessionId",
    },
  ],
]
`;

exports[`getTokens > when "idToken" has expired > resolves with a correct value 1`] = `"newDummyIdToken"`;

exports[`getTokens > when "refreshToken" has expired > rejects with a correct value 1`] = `[Error: Refresh token expired]`;
